src/app_engine/join_page_test.py

```diff
@@ -224,5 +225,124 @@ def testJoinDifferentRoomTypes(self):
 
+  def testAckRingingInvalidRoom(self):
+    self.addTestData()
+
+    # Room doesn't exist.
+    body = {
+        constants.PARAM_TYPE: constants.TYPE_RINGING,
+        constants.PARAM_CALLEE_GCM_ID: 'invalidCallee',
+        constants.PARAM_ROOM_ID: 'newRoom',
+    }
+    response = self.makePostRequest('/ack', json.dumps(body))
+    self.verifyResultCode(response, constants.RESPONSE_INVALID_ROOM)
+
+  def testAckRingingInvalidCallee(self):
+    self.addTestData()
+    # Callee doesn't exist.
+    room_id = 'callercallee'
+    self.requestCallAndVerify(room_id, 'caller1gcm1', 'callee1',
+                              constants.RESPONSE_SUCCESS)
+
+    body = {
+        constants.PARAM_TYPE: constants.TYPE_RINGING,
+        constants.PARAM_CALLEE_GCM_ID: 'invalidCallee',
+        constants.PARAM_ROOM_ID: room_id,
+    }
+
+    response = self.makePostRequest('/ack', json.dumps(body))
+    self.verifyResultCode(response, constants.RESPONSE_INVALID_CALLEE)
+
+  def testAckRingingWrongRoom(self):
+    self.addTestData()
+    room_id = 'callercallee'
+    self.requestCallAndVerify(room_id, 'caller1gcm1', 'callee1',
+                              constants.RESPONSE_SUCCESS)
+
+    # Callee not allowed in this room.
+    body = {
+        constants.PARAM_TYPE: constants.TYPE_RINGING,
+        constants.PARAM_CALLEE_GCM_ID: 'callee2gcm1',
+        constants.PARAM_ROOM_ID: room_id,
+    }
+
+    response = self.makePostRequest('/ack', json.dumps(body))
+    self.verifyResultCode(response, constants.RESPONSE_INVALID_CALLEE)
+
+  def testAckRinging(self):
+    self.addTestData()
+    room_id = 'callercallee'
+    self.requestCallAndVerify(room_id, 'caller1gcm1', 'callee1',
+                              constants.RESPONSE_SUCCESS)
+
+    body = {
+        constants.PARAM_TYPE: constants.TYPE_RINGING,
+        constants.PARAM_CALLEE_GCM_ID: 'callee1gcm1',
+        constants.PARAM_ROOM_ID: room_id,
+    }
+
+    self.clearGCMPayloads()
+    response = self.makePostRequest('/ack', json.dumps(body))
+    self.verifyResultCode(response, constants.RESPONSE_SUCCESS)
+
+    expected_payloads = [
+        self.createGCMRingingPayload(
+            ['caller1gcm1'],
+            room_id),
+    ]
+    self.verifyGCMPayloads(expected_payloads)
+
+    body = {
+        constants.PARAM_TYPE: constants.TYPE_RINGING,
+        constants.PARAM_CALLEE_GCM_ID: 'callee1gcm2',
+        constants.PARAM_ROOM_ID: room_id,
+    }
+
+    response = self.makePostRequest('/ack', json.dumps(body))
+    self.verifyResultCode(response, constants.RESPONSE_SUCCESS)
+    self.verifyGCMPayloads(expected_payloads)
+
+  def testAckRingingRoomFull(self):
+    self.addTestData()
+    room_id = 'callercallee'
+    self.requestCallAndVerify(room_id, 'caller1gcm1', 'callee1',
+                              constants.RESPONSE_SUCCESS)
+    self.requestAcceptAndVerify(room_id, 'callee1gcm1',
+                                constants.RESPONSE_SUCCESS)
+
+    # Room already full.
+    body = {
```

Comment:
Will do. I never think I'll have enough tests to justify doing it, and then all of a sudden I have a bunch of tests that look similar. I should learn my lesson and do it to start with.